### PR TITLE
based regex

### LINKED ISSUE #486: Remove more `Any` from `re` (`split`)

### LINKED ISSUE #487: special case the groups of literal patterns in regexp operations
```py
reveal_type(assert_not_none(re.match("a", input)).groups())  # str
reveal_type(assert_not_none(re.match("(a)", input)).groups())  # str, str
reveal_type(assert_not_none(re.match("(a)?", input)).groups())  # str, str | None
```

### PR DESCRIPTION
**Pull request by KotlinIsland**
- resolves #487
- resolves #486

### CODE REVIEW

File: docs/source/based_features.rst [hunk 1]
```diff
@@ -308,3 +308,18 @@ Checked Argument Names
...
+
+Regex Checks
+------------
+
+.. code-block:: python
+
+    re.compile("as(df")  #  error: missing ), unterminated subpattern at position 0  [regex]
+
+    if m := re.search("(a)?(b)", s):
+        reveal_type(m.groups()  # Revealed type is "(str | None, str)"
```
Comment (DetachHead) [inline docs/source/based_features.rst:321]:
```suggestion
        reveal_type(m.groups())  # Revealed type is "(str | None, str)"
```

File: mypy/erasetype.py [hunk 2]
```diff
@@ -199,7 +199,8 @@ class LastKnownValueEraser(TypeTranslator):
     def visit_instance(self, t: Instance) -> Type:
         if not t.last_known_value and not t.args:
             return t
-        return t.copy_modified(args=[a.accept(self) for a in t.args], last_known_value=None)
+        result = t.copy_modified(args=[a.accept(self) for a in t.args], last_known_value=None)
+        return result
```
Comment (DetachHead) [inline mypy/erasetype.py:203]: rem,ove

File: mypy/plugins/re.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+        groups = parse_groups(value, ctx.args[0][0], ctx.api.msg)
+        if groups is None:
+            return AnyType(TypeOfAny.from_error)
+    else:
+        assert isinstance(ctx.type, Instance)
+        groups = get_groups(ctx.type)
+        if groups is None:
+            return default
+    instance = default.items[0] if isinstance(default, UnionType) else default.args[0]
+    assert isinstance(instance, Instance)  # type: ignore[misc]
```
Comment (DetachHead) [inline mypy/plugins/re.py:47]: comment explaining why the type ignore is needed

File: mypy/plugins/re.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+def _parse_groups(value: str) -> Groups | re.error:
+    try:
+        p = re.compile(value)
+    except re.error as e:
+        return e
+
+    groups: list[int] = []
+    working = []
+    """
+    1 = capturing group
```
Comment (DetachHead) [inline mypy/plugins/re.py:271]: maybe clarify that the optional groups thing is  "the most important part"

File: mypy/plugins/re.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    0 = optional capturing group
+    -1 = non-capturing group
+    -2 = optional non-capturing group
```
Comment (DetachHead) [inline mypy/plugins/re.py:274]: document -3

File: mypy/plugins/re.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+                    working[d] = -3
+                union = -1
+            if depth == 0:
+                groups.extend(group for group in working if group > -1)
+                working.clear()
+    if union == 0:
+        for d in range(len(groups)):
+            if groups[d] == 1:
+                groups[d] = 0
+    assert len(groups) == p.groups
```
Comment (DetachHead) [inline mypy/plugins/re.py:353]: what happened?

File: mypy/plugins/re.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    ProperType,
+    TupleType,
+    Type,
+    TypedDictType,
+    TypeOfAny,
+    UnionType,
+    get_proper_type,
+)
+
+Groups = Tuple[Tuple[Optional[str], int], ...]
```
Comment (DetachHead) [inline mypy/plugins/re.py:28]:
```suggestion
Groups = Tuple[Tuple[Optional[str], bool], ...]
```

File: mypy/typeshed/stdlib/re.pyi [hunk 8]
```diff
@@ -133,25 +133,19 @@ class Pattern(Generic[AnyStr]):
     @overload
     def fullmatch(self, string: AnyStr, pos: int = 0, endpos: int = sys.maxsize) -> Match[AnyStr] | None: ...
     @overload
-    def split(self: Pattern[str], string: str, maxsplit: int = 0) -> list[str | Any]: ...
```
Comment (DetachHead) [inline mypy/typeshed/stdlib/re.pyi:136]: put comments that say these functions are intrinsically handled by mypy

File: test-data/unit/cmdline-based-regex.test [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+[file main.py]
+import re
+p1 = re.compile("()?()")
+s: str
+p2 = re.compile(s)
+if m := p2.match(""):
+    reveal_type(m.groups())
+if m := p1.match(""):
+    reveal_type(m.groups())
+[out]
```
Comment (DetachHead) [inline test-data/unit/cmdline-based-regex.test:202]: inline output

File: mypy/typeshed/stdlib/re.pyi (hunk 8 above)
Comment (KotlinIsland) [reply mypy/typeshed/stdlib/re.pyi:136]: check the diff

Comment (DetachHead) [reply mypy/typeshed/stdlib/re.pyi:136]: i think it should maybe have a comment on each one with special casing

File: mypy/plugins/re.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+                        working[d] = 0
+                union = -1
+            if depth == 0:
+                groups.extend(group for group in working if group > -1)
+                working.clear()
+    if union == 0:
+        for d in range(len(groups)):
+            if groups[d] == 1:
+                groups[d] = 0
+    assert len(groups) == p.groups, "parsed groups differ from compiled groups"
```
Comment (DetachHead) [inline mypy/plugins/re.py:352]: idk which is which but is it worth including the values for each one in the error message
```suggestion
    assert len(groups) == p.groups, f"parsed groups  ({groups}) differ from compiled groups ({p.groups})"
```

Comment (KotlinIsland) [reply mypy/plugins/re.py:352]: what values?

Comment (DetachHead) [reply mypy/plugins/re.py:352]: the group names

File: mypy/typeshed/stdlib/re.pyi (hunk 8 above)
Comment (KotlinIsland) [reply mypy/typeshed/stdlib/re.pyi:136]: Hows that look?

File: mypy/plugins/re.py (hunk 10 above)
Comment (KotlinIsland) [reply mypy/plugins/re.py:352]: yeah, it would have that already

**Review by DetachHead — APPROVED** [review]

**Pull request merged**